boards: kconfig: Add BOARD_QUALIFIERS

nordicjm opened the pull request:
Adds the user-provided board qualifiers to Kconfig so it can be used by application code

nordicjm committed:
boards: kconfig: Add BOARD_QUALIFIERS and BOARD_TARGET

Adds the user-provided board qualifiers to Kconfig so it can be
used by application code, along with one containing the full string

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

gmarull commented:
Note that looking at https://github.com/zephyrproject-rtos/zephyr/pull/69652

```c
#include <zephyr/sys/printk.h>

int main(void)
{
	printk("Hello world from %s%s\n", CONFIG_BOARD, CONFIG_BOARD_IDENTIFIER);

	return 0;
}
```

I'd also like to have a single Kconfig option with the board full identifier, code above looks a bit ugly.

nordicjm commented:
> Note that looking at #69652
> 
> ```c
> #include <zephyr/sys/printk.h>
> 
> int main(void)
> {
> 	printk("Hello world from %s%s\n", CONFIG_BOARD, CONFIG_BOARD_IDENTIFIER);
> 
> 	return 0;
> }
> ```
> 
> I'd also like to have a single Kconfig option with the board full identifier, code above looks a bit ugly.

Added

tejlmand commented on boards/Kconfig.v2:
```diff
@@ -15,4 +15,22 @@ config BOARD_$(BOARD_FULL_STRING)
+	string
+	default "$(BOARD)$(BOARD_IDENTIFIER)"
+	help
+	  Contains the full string board being.
```
The board being what ?

nordicjm replied:
Fixed

carlescufi commented on boards/Kconfig.v2:
```diff
@@ -15,4 +15,22 @@ config BOARD_$(BOARD_FULL_STRING)
+	  For example, if building for ``nrf5340dk/nrf5340/cpuapp`` then this will contain the
+	  value ``/nrf5340/cpuapp``.
+
+config BOARD_FULL_STRING
```
Didn't we use the words "fully-qualified identifier" for this?

carlescufi replied:
I see this is code that was already existing. Dismissing.

nashif commented:
I am getting:

```
CONFIG_BOARD_IDENTIFIER="/an385"
CONFIG_BOARD_FULL_STRING="mps2/an385"
```

nashif reviewed (changes requested):
CONFIG_BOARD_IDENTIFIER="/an385"
CONFIG_BOARD_FULL_STRING="mps2/an385"

seems to be wrong...

carlescufi commented:
> CONFIG_BOARD_IDENTIFIER="/an385" CONFIG_BOARD_FULL_STRING="mps2/an385"
> 
> seems to be wrong...

Why? As per the doc the identifier is what comes after the board name?

nashif commented:
> Why? As per the doc the identifier is what comes after the board name?

ok, need to do some catching up with the docs, by IMO this should not have a leading "/". Everywhere in the docs we are talking about things like `plank_defconfig / plank_<identifier>_defconfig`, the identifier does not have a "/", and tbh, when I look at this, I immediately suspect something is missing without looking at the docs. This is going to the extreme already, having identifiers start with a seperator string.

tejlmand commented:
> ok, need to do some catching up with the docs, by IMO this should not have a leading "/".

I thought you were referring to the content itself.

Internally in CMake and when passing `BOARD_IDENTIFIER` into Kconfig, it make sense to keep the `/`, as to not have to re-add it when concatenate info. Also when sanitizing strings.

But as a Kconfig string setting, which may be used for print output from code, then it could make sense to omit the leading `/`.

nordicjm commented:
> > ok, need to do some catching up with the docs, by IMO this should not have a leading "/".
> 
> I thought you were referring to the content itself.
> 
> Internally in CMake and when passing `BOARD_IDENTIFIER` into Kconfig, it make sense to keep the `/`, as to not have to re-add it when concatenate info. Also when sanitizing strings.
> 
> But as a Kconfig string setting, which may be used for print output from code, then it could make sense to omit the leading `/`.

Disagree, because a user should be able to combine CONFIG_BOARD and the variable to get the full output, it doesn't make sense to remove a character

tejlmand commented:
> Disagree, because a user should be able to combine CONFIG_BOARD and the variable to get the full output, it doesn't make sense to remove a character

that's exactly the problem with print strings like this.
Who is the user who will see this message and thus decide if those strings will be combined later ?

For example, in a build log then having the `/` makes sense as it is close to the build system (and thus useful when rebuilding)

In a sample like `hello_world`, it is debatable whether a `/` in front make sense or not.
I prefer the `/` in front for consistency with build system, but @nashif does bring up what I believe is a valid discussion.

nordicjm commented:
> > Disagree, because a user should be able to combine CONFIG_BOARD and the variable to get the full output, it doesn't make sense to remove a character
> 
> that's exactly the problem with print strings like this. Who is the user who will see this message and thus decide if those strings will be combined later ?
> 
> For example, in a build log then having the `/` makes sense as it is close to the build system (and thus useful when rebuilding)
> 
> In a sample like `hello_world`, it is debatable whether a `/` in front make sense or not. I prefer the `/` in front for consistency with build system, but @nashif does bring up what I believe is a valid discussion.

It doesn't make sense to not have it, that is the identifier that the user provided and that is the identifier of the board which can be combined with the board name, it wouldn't make much sense to output `nrf5340dknrf5340/cpuapp`, that just looks silly

nashif commented:
The docs is actually clear about what we call identifier:

> "A / is used as separator between the board name and the following: SoC, CPU cluster, and variant identifiers."

So, adding "/" to be part of the identifier does not make any sense and is plain WRONG.

nordicjm commented:
> The docs is actually clear about what we call identifier:
> 
> > "A / is used as separator between the board name and the following: SoC, CPU cluster, and variant identifiers."
> 
> So, adding "/" to be part of the identifier does not make any sense and is plain WRONG.

Have you looked at the text and examples on https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-and-board-identifiers and https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#boards-socs-etc because they all show / as being part of the identifier and explain it is part of it, they say the slash is the separation character because it is, but it's still part of the identifier

nordicjm committed:
samples: hello_world: Use full board target string for hwmv2 boards

Outputs the full board target string (including qualifiers) when
using a hwmv2 board

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

gmarull reviewed (commented):
can we just introduce `CONFIG_BOARD_FULL_STRING`, and leave `BOARD_IDENTIFIER` for a future discussion? It seems to be the one we need in practice right now.

carlescufi commented:
> > The docs is actually clear about what we call identifier:
> > > "A / is used as separator between the board name and the following: SoC, CPU cluster, and variant identifiers."
> > 
> > 
> > So, adding "/" to be part of the identifier does not make any sense and is plain WRONG.
> 
> Have you looked at the text and examples on https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-and-board-identifiers and https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#boards-socs-etc because they all show / as being part of the identifier and explain it is part of it, they say the slash is the separation character because it is, but it's still part of the identifier

@nordicjm this is actually being changed here: https://github.com/zephyrproject-rtos/zephyr/pull/69802/files

nordicjm committed:
scripts: kconfig: Add substring function

Adds a substring function

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

gmarull commented on samples/hello_world/src/main.c:
```diff
@@ -8,6 +8,13 @@
 int main(void)
 {
-	printf("Hello World! %s\n", CONFIG_BOARD);
+	printf("Hello World! %s\n",
+#ifdef CONFIG_BOARD_TARGET
+	       CONFIG_BOARD_TARGET
+#else
+	       CONFIG_BOARD
+#endif
+	);
```
```suggestion
	printf("Hello World! %s\n", CONFIG_BOARD_TARGET);
```

let's not make hello world uglier.

nordicjm replied:
Not possible because it doesn't exist for hwmv1 devices, which are still supported

gmarull replied:
then I'd define `CONFIG_BOARD_TARGET` to the equivalent when in HWMv1

nordicjm replied:
No, we should not be adding hwmv2 defines to hwmv1 boards

gmarull replied:
then we need a solution

nordicjm replied:
Well I mean it doesn't considering the term was coined in the decision that concluded on Friday, it was simple a "board" in hwmv1

gmarull replied:
then why not insert BOARD_TARGET = BOARD in v1 Kconfig? this will make life easier in all samples relying on BOARD_TARGET.

nordicjm replied:
Because it's a hwmv2 concept not a hwmv1 concept and can be used from application code to detect a hwmv2 board

gmarull replied:
> can be used from application code to detect a hwmv2 board

where is this needed?

nordicjm replied:
Anywhere?

nordicjm replied:
It's not passed in C to the application which is where this is

nordicjm replied:
> `CONFIG_BOARD_TARGET` should always be defined.
> 
> Although we have discussed terminology in context of HWMv2, then I would argue that for HWMv1 we should have `CONFIG_BOARD_TARGET` == `CONFIG_BOARD` simply because HWMv1 doesn't have the concept of qualifiers.

I do not agree, hwmv1 is now end of life, it doesn't make sense to introduce hwmv2 with terms **specifically for hwmv2** then just apply them to hwmv1 for releases 3.7 and 4.0 when it's not in any of the previous released when we all agreed that hwmv2 is now the standard

gmarull replied:
we simply need `CONFIG_BOARD_TARGET` defined everywhere, forcing people to use ifdeffery just for an obsolete hw model doesn't make any sense.

nordicjm replied:
> we simply need `CONFIG_BOARD_TARGET` defined everywhere, forcing people to use ifdeffery just for an obsolete hw model doesn't make any sense.

Adding features to a deprecated and unsupported hardware model where the terms were never applied doesn't make any sense

tejlmand replied:
> Adding features to a deprecated and unsupported hardware model where the terms were never applied doesn't make any sense

This is about being pragmatic.
Comparing the two options of:
Adding `CONFIG_BOARD_TARGET` one place vs having `#ifdef CONFIG_BOARD_TARGET` n-places.

then I choose the first, "Adding `CONFIG_BOARD_TARGET` one place".

nordicjm replied:
>  Adding CONFIG_BOARD_TARGET one place vs having #ifdef CONFIG_BOARD_TARGET n-places.

We're not adding it to multiple places, we're adding it to hello world, and that's it

gmarull replied:
> > Adding CONFIG_BOARD_TARGET one place vs having #ifdef CONFIG_BOARD_TARGET n-places.
> 
> We're not adding it to multiple places, we're adding it to hello world, and that's it

Other samples (eg multitarget hello world, mbox, IPC) which typically run on multiple cores will also need this to become more user-friendly. Spreading such ifdeffery everywhere for no added value doesn't seem like a good idea.

tejlmand replied:
> We're not adding it to multiple places, we're adding it to hello world, and that's it

No, any place which would want to print `CONFIG_BOARD_TARGET` would need to do this.
So might be one place in this particular PR, but would end up in n-places later.

tejlmand replied:
it could be in cases like this where knowing the CPU cluster on a multicore SoC would be nice:
https://github.com/zephyrproject-rtos/zephyr/blob/42c8b7fb00a8f6343966e140b4bba593bb92cfdb/tests/boot/test_mcuboot/src/main.c#L14

or when printing board in `userspace/hello_world`, then why should that printing deviate from simple `hello_world` ?
https://github.com/zephyrproject-rtos/zephyr/blob/42c8b7fb00a8f6343966e140b4bba593bb92cfdb/samples/userspace/hello_world_user/src/main.c#L20-L22

gmarull reviewed (changes requested):
-1 on making hello world uglier, needs a better solution

tejlmand commented on boards/Kconfig.v2:
```diff
@@ -15,4 +16,23 @@ config BOARD_$(BOARD_FULL_STRING)
+	  For example, if building for ``nrf5340dk/nrf5340/cpuapp`` then this will contain the
+	  value ``nrf5340/cpuapp``.
+
+config BOARD_TARGET
+	string
+	default "$(BOARD)$(BOARD_IDENTIFIER)"
```
This is not `BOARD_TARGET` according to 3.a here: https://github.com/zephyrproject-rtos/zephyr/issues/69905

It lacks the revision field.

nordicjm replied:
Then how can this be done in Kconfig given that BOARD_REVISION might not exist and if it does exist there is no @ symbol?

nordicjm replied:
Fixed

tejlmand commented on boards/Kconfig.v2:
```diff
@@ -1,9 +1,10 @@
 
 BOARD_STRING := $(sanitize_upper,$(BOARD))
 BOARD_FULL_STRING := $(sanitize_upper,$(BOARD)$(BOARD_IDENTIFIER))
+BOARD_QUALIFIERS_WITHOUT_SEPARATOR := $(substring,$(BOARD_IDENTIFIER),1)
```
afaiu the board qualifiers is without the first separator.

```suggestion
BOARD_QUALIFIERS := $(substring,$(BOARD_IDENTIFIER),1)
```

We anyway need to update the CMake code according to #69905 3.a, and then we should simply just pass in the qualifiers as it has new been defined by the terminology discussion.

nordicjm replied:
Fixed

nordicjm commented:
Moved commit of contention to https://github.com/zephyrproject-rtos/zephyr/pull/70232

gmarull commented:
> Moved commit of contention to #70232

it's not really a commit, because without handling hwmv1 here, all samples using CONFIG_BOARD_TARGET unconditionally (what is being requested) won't compiler there.

nordicjm commented:
> > Moved commit of contention to #70232
> 
> it's not really a commit, because without handling hwmv1 here, all samples using CONFIG_BOARD_TARGET unconditionally (what is being requested) won't compile there.

No in-tree sample is using it unconditionally

gmarull commented:
> > > Moved commit of contention to #70232
> > 
> > 
> > it's not really a commit, because without handling hwmv1 here, all samples using CONFIG_BOARD_TARGET unconditionally (what is being requested) won't compile there.
> 
> No in-tree sample is using it unconditionally

yes, but that's the goal, so the problem is not solved.

tejlmand commented:
> Moved commit of contention to #70232

Problem with that is that we loose the discussion history and arguments in favor / against the solution.

fabiobaltieri commented:
are these two needed now? ithe problem is just the hello world string maybe it can be changed once hwmv1 support is dropped

tejlmand commented on boards/Kconfig.v2:
```diff
@@ -15,4 +16,24 @@ config BOARD_$(BOARD_FULL_STRING)
+	  For example, if building for ``nrf5340dk/nrf5340/cpuapp`` then this will contain the
+	  value ``nrf5340/cpuapp``.
+
+config BOARD_TARGET
+	string
+	default "$(BOARD)@$(BOARD_REVISION)$(BOARD_IDENTIFIER)" if "$(BOARD_REVISION)" != ""
+	default "$(BOARD)$(BOARD_IDENTIFIER)"
+	help
+	  Contains the board target (full string including name, revision, soc, cluster and
+	  variant) of the board being used.
+
+	  For example, if building for ``nrf5340dk/nrf5340/cpuapp`` then this will contain the
+	  value ``nrf5340dk/nrf5340/cpuapp``.
```
Please make this generally available together with other `config BOARD` in this Kconfig file:
https://github.com/zephyrproject-rtos/zephyr/blob/4b8d3dbed8a19a28a37b3fb5b46787ddb14d98b0/boards/Kconfig#L3-L21

tejlmand replied:
Note that the initial reason for the `Kconfig.v1` and `Kconfig.v2` split is because v1 boards (and SoCs) are using `choice` selections which prevents boards from using select and instead relies on a `CONFIG_BOARD_PLANK=y` and `CONFIG_SOC_FOO=y` in the `_defconfig`.

Therefore we needed to split into two trees, to remove the choice.

**Not** because we needed to  separated new Kconfigs from old Kconfigs.

See also: https://github.com/zephyrproject-rtos/zephyr/blob/b839f411613e500a8d20592d40709d1317feadca/boards/Kconfig.v1#L5-L6
https://github.com/zephyrproject-rtos/zephyr/blob/b839f411613e500a8d20592d40709d1317feadca/soc/Kconfig.v1.choice#L5-L6

tejlmand commented on boards/Kconfig.v2:
```diff
@@ -15,4 +16,13 @@ config BOARD_$(BOARD_FULL_STRING)
 	help
 	  Kconfig symbol identifying the board including full board identifier.
 
+config BOARD_QUALIFIERS
```
technically, the `BOARD_QUALIFIERS` is not required to be in this file, it could go directly to `<boards>/Kconfig`, and would be empty for a v1 board.

One reason the `BOARD_$(BOARD_STRING)` and `BOARD_$(BOARD_FULL_STRING)` must be in Kconfig.v2 is because those may collide with `config BOARD_<name>` names used for v1 boards, as those used to be defined in `<board-dir>/Kconfig.board` and exists inside the choice.

For example like in:
https://github.com/zephyrproject-rtos/zephyr/blob/468eb56cf242eedba62006ee758700ee6148763f/boards/arm/nrf9160dk_nrf9160/Kconfig.board#L8
https://github.com/zephyrproject-rtos/zephyr/blob/468eb56cf242eedba62006ee758700ee6148763f/boards/arm/nrf9160dk_nrf9160/Kconfig.board#L11

But the qualifier can be moved to `<boards>/Kconfig` later when support for v1 is removed.

nashif commented on boards/Kconfig:
```diff
@@ -19,6 +19,17 @@ config BOARD_REVISION
 
+config BOARD_TARGET
+	string
+	default "$(BOARD)@$(BOARD_REVISION)$(BOARD_IDENTIFIER)" if "$(BOARD_REVISION)" != ""
```
It is confusing that we are still using the BOARD identifier terminology here.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/180017/37b9cd60-f621-4012-9f99-4e24e0f0cc2a)

henrikbrixandersen replied:
Agreed.

gmarull replied:
maybe deprecate `CONFIG_BOARD` and introduce `CONFIG_BOARD_NAME`?

nordicjm replied:
They will be updated in a separate PR since there's usage through the whole zephyr tree of it as per https://github.com/zephyrproject-rtos/zephyr/pull/69904#issuecomment-2003607242

carlescufi replied:
I agree, we should fix this even though I still haven't finished the work documenting this in https://github.com/zephyrproject-rtos/zephyr/pull/69802

nashif replied:
more confusion with links to multiple PRs, all of which talk about different terminology, even the documentation change one still talks about identifiers. Can we do this in order somehow, right now it is impossible to know what is an oversight and what is there just for backward compatibility etc and what the next steps are.

tejlmand replied:
created https://github.com/zephyrproject-rtos/zephyr/pull/70438.

If we merge that one first, then we can rebase all PRs accordingly and use the new terminology everywhere.

carlescufi commented:
Waiting for the hotfix here: https://github.com/zephyrproject-rtos/zephyr/pull/70565
